js/libs/core.js

```diff
@@ -67,6 +67,77 @@ Fliplet.FormBuilder = (function() {
         component.methods = {};
       }
 
+      component.methods.setSourceValue = function(data) {
+        var result;
+        var $vm = this;
+
+        switch (data.source) {
+          case 'profile':
+            if (!data.key) {
+              throw new Error('A key is required to fetch data from the user\'s profile');
+            }
+
+            result = Fliplet.User.getCachedSession()
+              .then(function(session) {
+                if (session && session.entries) {
+                  if (session.entries.dataSource) {
+                    return session.entries.dataSource.data[data.key];
+                  }
+
+                  if (session.entries.saml2) {
+                    return session.entries.saml2.data[data.key];
+                  }
+
+                  if (session.entries.flipletLogin) {
+                    return session.entries.flipletLogin.data[data.key];
+                  }
+                }
+
+                return Fliplet.Profile.get(data.key);
+              })
+            break;
+          case 'query':
+            if (!data.key) {
+              throw new Error('A key is required to fetch data from the navigation query parameters');
+            }
+
+            result = Fliplet.Navigate.query[data.key];
+            break;
+          case 'appStorage':
+            if (!data.key) {
+              throw new Error('A key is required to fetch data from app storage');
+            }
+
+            result = Fliplet.App.Storage.get(data.key);
+            break;
+          default:
+            result = this.value;
+        }
+
+        if (!(result instanceof Promise)) {
+          result = Promise.resolve(result);
+        }
+
+        return result.then(function (value) {
+          if (typeof value === 'undefined') {
+            value = '';
```

Comment:
If the value is `undefined`, what's wrong with returning `undefined` as it is?

Reply:
@tonytlwu for the most field I've worked with it is nothing, but it was taken from the original `set` function at this [line](https://github.com/Fliplet/fliplet-widget-form-builder/blob/a9ec4ffb962c5867676ce99046b53363ea467586/js/libs/form.js#L848) and I thought it is needed for some fields I haven't worked with.
Should I remove it?

Reply:
@squallstar What's your opinion? [You coded it](https://github.com/Fliplet/fliplet-widget-form-builder/blame/a9ec4ffb962c5867676ce99046b53363ea467586/js/libs/form.js#L848).

Reply:
@tonytlwu it may have been due to wanting to set the value to an object that has a common set of instance methods such as `indexOf`. The default value for an input field (at least a text one) is an empty string. If such field is accessing string instance methods then it would fail if we set that to undefined

Reply:
OK. Let's keep the enforced `""` then. @YaroslavOvdii Good job for finding this from `.set()` to keep our consistency.